File: samples/gpu/openmp/example1.cpp
/*
Very simple example of openmp offloading to GPU adopted from
https://gist.github.com/anjohan/9ee746295ea1a00d9ca69415f40fafc9

To compile on GPU box:
No omp
/home/Niki.Zadeh/opt/llvm_14.0.5_gcc11/install/bin/clang++ example1.cpp -o example1_cl  -L/opt/gcc/11.3.0/lib64  ; time ./example1_cl
 size        runtime sum
 100000000  251.970   687.194763183593750
real	4m12.370s
user	4m11.875s

OMP No offload
/home/Niki.Zadeh/opt/llvm_14.0.5_gcc11/install/bin/clang++ example1.cpp -o example1_cl_omp -fopenmp  -L/opt/gcc/11.3.0/lib64  ; time ./example1_cl_omp
 size        runtime sum
 100000000   34.720   687.194763183593750
real	0m35.040s
user	9m15.786s

OMP Offload to GPU
//clang++ requires [0:N], otherwise errors out
/home/Niki.Zadeh/opt/llvm_14.0.5_gcc11/install/bin/clang++ example1.cpp -o example1_cl_omp_target -fopenmp -fopenmp-targets=nvptx64  -L/opt/gcc/11.3.0/lib64  ; time ./example1_cl_omp_target
clang-14: warning: CUDA version is newer than the latest supported version 11.5 [-Wunknown-cuda-version]
 size        runtime sum
 100000000    5.062   687.194763183593750

real	0m7.270s
user	0m3.164s

nvc++   example1.cpp -o example1_nvc ;time ./example1_nvc
 size        runtime sum
 100000000  422.520   687.194763183593750
real	7m2.776s

nvc++ -mp  example1.cpp -o example1_nvc_mp ;time ./example1_nvc_mp
 size        runtime sum
 100000000   54.562   687.194763183593750
real	0m54.912s

nvc++ -mp -ta=tesla  example1.cpp -o example1_nvc_target ;time ./example1_nvc_target
 size        runtime sum
 100000000   50.846   687.194763183593750
real	0m51.015s

****If 3. replaced by 3 in y[i] += 3.*x[i]/N; 2x ********
nvc++ -mp -ta=tesla  example1.cpp -o example1_nvc_target ;time ./example1_nvc_target
 size        runtime sum
 100000000   25.085   687.194763183593750
real	0m25.238s


 */
#include <stdio.h>
#include <time.h>
#include <omp.h>

int main(){
  int N = 1e8;
  float *x = new float[N];
  float *y = new float[N];
  double runtime;
  //clock_t before = clock();
  runtime = omp_get_wtime();
#pragma omp target teams distribute parallel for map(tofrom: x[0:N], y[0:N]) 
  for(int i = 0; i < N; i++){
    x[i] = i;
    for(int j = 0; j < 1000; j++){
      y[i] += 3*x[i]/N;
    }
  }
  runtime = omp_get_wtime() - runtime;//End timer
  //clock_t difference = clock() - before;
  //float runtime = difference * 1000 / CLOCKS_PER_SEC/1000.f;
  
  float sum=0.0;
  for(int i = 0; i < N; i++){
    sum += y[i];    
  }

  printf(" size        runtime sum\n");
  printf("%10d %8.3f %21.15f\n", N, runtime, sum/N);

  delete [] x;
  delete [] y;
}

File: samples/gpu/openmp/example1.f90
!
!Very simple example of openmp offloading to GPU adopted from
!https://gist.github.com/anjohan/9ee746295ea1a00d9ca69415f40fafc9

!To compile on GPU box:
!No omp
!\rm ./example1_fl ; /home/Niki.Zadeh/opt/llvm/install/bin/flang example1.f90 -o example1_fl  -L/opt/gcc/11.3.0/lib64 ; time ./example1_fl
! size        runtime sum
! 100000000   0.000  687.194763183593750
!real	5m47.690s
!user	5m47.332s
!
!nvfortran  example1.f90 -o example1_nvf ; time ./example1_nvf
! size        runtime sum                omp_num_threads
! 100000000   0.000  687.194763183593750   0
!real	4m5.391s
!user	4m5.226s
!
!-O3
!/home/Niki.Zadeh/opt/llvm/install/bin/flang example1.f90 -O3 -L/opt/gcc/11.3.0/lib64 -o example1_fl ; time ./example1_fl
! size        runtime sum                omp_num_threads
! 100000000   0.000  687.194763183593750****
!real	0m20.021s
!user	0m19.890s
!
!gfortran -O3 example1.f90 -o example1_gf; time ./example1_gf
! size        runtime sum                omp_num_threads
! 100000000   0.000  687.194763183593750****
!real	0m20.037s
!user	0m19.903s
!
!-fopenmp
!/home/Niki.Zadeh/opt/llvm/install/bin/flang example1.f90 -O3 -L/opt/gcc/11.3.0/lib64 -fopenmp -o example1_fl_omp ; time ./example1_fl_omp
! size        runtime sum                omp_num_threads
! 100000000  24.786  687.194763183593750   1
!real	0m25.071s
!user	6m31.081s
!
!gfortran -O3 example1.f90 -fopenmp -o example1_gf_omp; time ./example1_gf_omp
! size        runtime sum                omp_num_threads
! 100000000  24.755  687.194763183593750   1
!real	0m24.903s
!user	6m30.967s
!
!nvfortran  example1.f90 -o example1_nvf ; time ./example1_nvf
! size        runtime sum                omp_num_threads
! 100000000   0.000  687.194763183593750   0
!real	4m5.391s
!user	4m5.226s
!
!nvfortran  example1.f90 -O3 -o example1_nvf ; time ./example1_nvf
! size        runtime sum                omp_num_threads
! 100000000   0.000 1505.299316406250000****         !!!!!-O3 does it wrong
!real	0m4.871s
!user	0m4.301s
!
!nvfortran -Mpreprocess  -ta=tesla,cc60 -mp  example1.f90 -o example1_nvf ;time ./example1_nvf
! size        runtime sum                omp_num_threads
! 100000000   5.812  687.194763183593750   1
!real	0m6.545s
!user	1m30.543s
!
!module load cuda/11.7
!module load nvhpc-no-mpi/22.5
!nvfortran -Mpreprocess  -ta=tesla,cc60,cuda11.7 -mp  example1.f90 -o example1_nvf_target ;time ./example1_nvf_target
! size        runtime sum                omp_num_threads
! 100000000   5.599  687.194763183593750   1            !!!GPU idle!!!
!real	0m6.116s
!user	1m28.413s
!
!nvfortran -Mpreprocess  -ta=tesla,cc70,cuda11.7 -mp=gpu  example1.f90 -o example1_nvf_target ;time ./example1_nvf_target
! size        runtime sum                omp_num_threads
! 100000000   0.544  687.194763183593750   1
!real	0m0.920s
!
!
!ACC
!pgf90 -acc -ta=nvidia:managed -Minfo=accel -o example1_acc_managed example1.f90; time ./example1_acc_managed
! size        runtime sum                omp_num_threads
! 100000000   0.000  687.194763183593750****
!real	0m1.261s
!user	0m0.491s
!
!DO CONCURRENT with GPU offload (Note -stdpar and comment out $omp target 
!nvfortran -mp -stdpar  example1.f90 -o example1_nvf_docon ;time ./example1_nvf_docon 
! size        runtime sum                omp_num_threads
! 100000000   0.237  687.194763183593750   1

program example1
  implicit none
  integer,parameter :: N = 1e8
  integer :: i,j
  real, dimension (:),  allocatable :: x,y
  real*8 :: run_time, omp_get_wtime
  integer :: nthread, OMP_GET_NUM_THREADS,OMP_GET_THREAD_NUM
  real :: sum

  allocate ( x(0:N-1) , y(0:N-1) ) ; x=0.0; y=0.0
!$   nthread = omp_get_num_threads()
!$   run_time = omp_get_wtime() 
!NOacc data copyin(x) copyout(y)
!NOacc kernels present(x,y)
!Nomp target teams distribute parallel do map(tofrom: x, y)
!  do i=0,N-1
  do concurrent(i=0:N-1)
    x(i) = i*1.
    do j=1,1000
      y(i) = y(i)+ 3.*x(i)/N
    enddo
  enddo
!NOacc end kernels
!NOacc end data

!$   run_time = omp_get_wtime() - run_time;
  
  sum=0.0
  do i=0,N-1; sum = sum+y(i); enddo

  write(*,'(a)') ' size        runtime sum                omp_num_threads'
  write(*,'(i10,f8.3,f21.15, i4)') N, run_time, sum/N,nthread 

  deallocate(x,y)
end program example1

